tests/gtest/dataset_test.cpp

```diff
@@ -0,0 +1,88 @@
+ *     notice, this list of conditions and the following disclaimer.
+ *  2. Redistributions in binary form must reproduce the above copyright
+ *     notice, this list of conditions and the following disclaimer in the
+ *     documentation and/or other materials provided with the distribution.
+ *  3. Neither the name of the copyright holder nor the
+ *     names of its contributors may be used to endorse or promote products
+ *     derived from this software without specific prior written permission.
+ *
+ *  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ *  AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ *  IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ *  ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ *  LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ *  CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ *  SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ *  INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ *  CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ *  ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include <gmock/gmock.h>
+#include <gtest/gtest.h>
+
+#include <openthread/border_agent.h>
+#include <openthread/dataset.h>
+#include <openthread/dataset_ftd.h>
+#include <openthread/instance.h>
+#include <openthread/ip6.h>
+#include <openthread/thread.h>
+#include <openthread/platform/time.h>
+#include "gmock/gmock.h"
+
+#include "fake_platform.hpp"
+
+using namespace ot;
+using ::testing::AnyNumber;
+using ::testing::AtLeast;
+using ::testing::MockFunction;
+using ::testing::Truly;
+
+template <typename R, typename... A> class MockCallback : public testing::MockFunction<R(A...)>
+{
+public:
+    static R CallWithContext(A... aArgs, void *aContext)
+    {
+        return static_cast<MockCallback *>(aContext)->Call(aArgs...);
+    };
+};
+
+TEST(otDatasetSetActiveTlvs, shouldTriggerStateCallbackOnSuccess)
+{
+    FakePlatform fakePlatform;
+
+    typedef MockCallback<void, otChangedFlags> MockStateCallback;
+
+    MockStateCallback mockStateCallback;
+
+    EXPECT_CALL(mockStateCallback, Call).Times(AnyNumber());
+
+    EXPECT_CALL(mockStateCallback, Call(Truly([](otChangedFlags aChangedFlags) -> bool {
+                    return (aChangedFlags & OT_CHANGED_ACTIVE_DATASET);
+                })))
+        .Times(AtLeast(1));
+
+    otError error = otSetStateChangedCallback(FakePlatform::CurrentInstance(), MockStateCallback::CallWithContext,
+                                              &mockStateCallback);
+    assert(error == OT_ERROR_NONE);
+
+    otOperationalDataset     dataset;
+    otOperationalDatasetTlvs datasetTlvs;
+
+    error = otDatasetCreateNewNetwork(FakePlatform::CurrentInstance(), &dataset);
```

Comment:
Yakun,

My understanding is that this PR provides a fake (virtual) OpenThread platform implementation and then allows tests to be written to override a subset of the platform APIs and write tests to validate that the platform APIs are properly called.

**The question is whether such a test would be better under "unit tests" and if we need this new model?**

Unit tests already provide similar behavior. They provide an empty **weak** implementation of all platform APIs and allow each test to provide its own implementation, overriding the weak one if the test wants to check something.

We already have unit tests that use this model:
- `test_timer` provides a special implementation of `otPlatAlarm` APIs and tests timer behavior, particularly how the platform APIs are called.
- `test_routing_manager` does something similar with all `otPlatInfraIf` APIs (validating they are called properly and checking the passed-in RA, RS, or ... messages sent by OpenThread code).
- `test_srp_server` does the same with hooks to the "advertising proxy" APIs, validating different interactions between the OpenThread core and the platform (e.g., late callbacks, ignored callbacks, etc.).
- `test_mdns` does the same with its APIs.
- ...

The example `dataset_test.cpp` in this PR can easily be written as a unit test. 
- Start an instance. 
- Call `otSetStateChangedCallback()` to set a callback provided by the test itself (which tracks if it's called and saves the event flags in a local variable/array).
- Invoke `otDatasetCreateNewNetwork()`.
- Validate that the callback was called with the proper flags.

I see clear advantages to the unit test model:
- It's an already existing model providing a similar function.
- It's simpler to understand and follow the test code and steps.
- The new model in this PR requires people to understand and learn new commands and formatting. For example, this block is not easy to follow (requires familiarity with all these gtest macros/functions):

```c++
EXPECT_CALL(mockStateCallback, Call).Times(AnyNumber());

EXPECT_CALL(mockStateCallback, Call(Truly([](otChangedFlags aChangedFlags) -> bool {
           return (aChangedFlags & OT_CHANGED_ACTIVE_DATASET);
       })))
.Times(AtLeast(1));
```

Reply:
> Yakun,
> 
> My understanding is that this PR provides a fake (virtual) OpenThread platform implementation and then allows tests to be written to override a subset of the platform APIs and write tests to validate that the platform APIs are properly called.
> 
> **The question is whether such a test would be better under "unit tests" and if we need this new model?**
> 
> Unit tests already provide similar behavior. They provide an empty **weak** implementation of all platform APIs and allow each test to provide its own implementation, overriding the weak one if the test wants to check something.
> 
> We already have unit tests that use this model:
> 
> * `test_timer` provides a special implementation of `otPlatAlarm` APIs and tests timer behavior, particularly how the platform APIs are called.
> * `test_routing_manager` does something similar with all `otPlatInfraIf` APIs (validating they are called properly and checking the passed-in RA, RS, or ... messages sent by OpenThread code).
> * `test_srp_server` does the same with hooks to the "advertising proxy" APIs, validating different interactions between the OpenThread core and the platform (e.g., late callbacks, ignored callbacks, etc.).
> * `test_mdns` does the same with its APIs.
> * ...
> 
> The example `dataset_test.cpp` in this PR can easily be written as a unit test.
> 
> * Start an instance.
> * Call `otSetStateChangedCallback()` to set a callback provided by the test itself (which tracks if it's called and saves the event flags in a local variable/array).
> * Invoke `otDatasetCreateNewNetwork()`.
> * Validate that the callback was called with the proper flags.
> 
> I see clear advantages to the unit test model:
> 
> * It's an already existing model providing a similar function.
> * It's simpler to understand and follow the test code and steps.
> * The new model in this PR requires people to understand and learn new commands and formatting. For example, this block is not easy to follow (requires familiarity with all these gtest macros/functions):
> 
> ```c++
> EXPECT_CALL(mockStateCallback, Call).Times(AnyNumber());
> 
> EXPECT_CALL(mockStateCallback, Call(Truly([](otChangedFlags aChangedFlags) -> bool {
>            return (aChangedFlags & OT_CHANGED_ACTIVE_DATASET);
>        })))
> .Times(AtLeast(1));
> ```

Thanks for raising the concern.

I understand that the current unit tests leverage weak symbols for overrides, which is functional. However, introducing a fake platform and gtest doesn't enable something previously impossible. Instead, **it significantly improves the efficiency of writing unit/component tests, making them more readable, maintainable, and extensible**. The dataset test example is a basic illustration of how the new model can be used.

**Leveraging a fake platform allows us to create a platform tailored to the specific needs of a single test or test step**. This means we can easily have dedicated mocks for various cases within a test unit, making each test case self-contained and leading to better organized tests. For example, consider a scenario where we have two test cases in a unit, each requiring different overrides for the same platform APIs. With the current model, we'd either have to handle these differences within a single overridden platform API or create dedicated code in the platform API override that supports switching behaviors within each test case. The fake platform addresses this by allowing each test case to define its platform's own behavior.

Furthermore, gtest simplifies the process of recording and inspecting function calls with varying parameters, return values, and timing. This allows us to focus on verifying expected behaviors rather than dealing with common requirements for each test case not only for platform APIs, but also for user APIs.

Regarding the new macros/functions, [gtest](https://github.com/google/googletest) is a popular open source test framework released by Google. The *commands* and *formatting* are widely used beyond Google. Some of them are very self explained, some are similar to existing terms in OpenThread, while some are new. I admit it may require some people to get familiar with them. Luckily, they all have [good documentation](http://google.github.io/googletest/) so it may not be that hard to pick them up. On the other hand, with the unified syntax, I expect it will help improve the readability in the long term.

@abtink What do you think?

Reply:
> **it significantly improves the efficiency of writing unit/component tests, making them more readable, maintainable, and extensible.**

I don't agree with the above points about the new fake platform (and gtest). 

* **Harder to write tests:** Developers need to first learn the new syntax and related commands, increasing the barrier to entry. 
* **Reduced readability:** The code is not more readable as it requires understanding the new syntax and commands. Even for developers familiar with it, the syntax can be a bit dense and difficult to parse quickly.  Do you really find this code easy to follow? 
   ```c++
   EXPECT_CALL(mockStateCallback, Call).Times(AnyNumber());

   EXPECT_CALL(mockStateCallback, Call(Truly([](otChangedFlags aChangedFlags) -> bool {
           return (aChangedFlags & OT_CHANGED_ACTIVE_DATASET);
       })))
   .Times(AtLeast(1));
   ``` 
* **Less maintainable:**  Adding a dependency on a new tool/library (gtest) makes it less maintainable, even if we argue that gtest is well-supported. 
* **Not necessarily more extensible:** I think the unit test model allows for the same or maybe even more flexibility and extensibility when writing tests. The variety of tests we already have is a good indicator of this.

Reply:
As we discussed offline, I'll add additional complicated test cases based on the fake platform in separate PR after this one. I'll resolve this comment for now. Please feel free to re-open if you have other concerns. Thanks again for sharing the these concerns.